README.md
[![Build Status](https://travis-ci.org/MTG/SymbTr-extras.svg?branch=master)](https://travis-ci.org/MTG/SymbTr-extras) [![Code Climate](https://codeclimate.com/github/MTG/SymbTr-extras/badges/gpa.svg)](https://codeclimate.com/github/MTG/SymbTr-extras) [![codecov.io](https://codecov.io/github/MTG/SymbTr-extras/coverage.svg?branch=master)](https://codecov.io/github/MTG/SymbTr-extras?branch=master) [![DOI](https://zenodo.org/badge/52790094.svg)](https://zenodo.org/badge/latestdoi/52790094)

SymbTr-extras
===========
Basic tools to manipulate the SymbTr-scores 

Introduction
------------
This repository hosts a set of tools to validate and manipulate the SymbTr scores and get basic statistics from them. The tools are mostly used for maintaining the [SymbTr repository](https://github.com/MTG/SymbTr) and keeping a track of the stored score metadata and attributes.

If you are using these tools in your work, please cite the dissertation:

> Şentürk, S. (2016). [Computational Analysis of Audio Recordings and Music Scores for the Description and Discovery of Ottoman-Turkish Makam Music](http://sertansenturk.com/research/works/phd-thesis/). PhD thesis, Universitat Pompeu Fabra, Barcelona, Spain.

Usage
------------
Please refer to the interactive demos [apply_extras_txt_file.ipynb](https://github.com/MTG/SymbTr-extras/blob/master/apply_extras_txt_file.ipynb) and [change_symbtr_filename](https://github.com/MTG/SymbTr-extras/blob/master/change_symbtr_filename.ipynb) for the basic usage. To process many documents you can use the [jupyter notebooks](http://jupyter.org/) in the [wrapper](https://github.com/MTG/SymbTr-extras/tree/master/wrappers) folder.

This repository is also used as a submodule in the [SymbTr collection](https://github.com/MTG/SymbTr). For this reason, the paths in the jupyter notebooks are given according to the relative path of the submodule. If you want to use these notebooks outside the submodule, you should change these paths accordingly.

Installation
------------
If you want to install the package, it is recommended to install it and all the dependencies into a virtualenv. In the terminal, do the following:

    virtualenv env
    source env/bin/activate
    python setup.py install

If you want to be able to edit files and have the changes be reflected, then
install the repository like this instead:

    pip install -e .

Now you can install the rest of the dependencies:

    pip install -r requirements

Authors
-------
Sertan Senturk
contact@sertansenturk.com


Changelog.md
#### SymbTr-extras v0.4.0
 - Added a wrapper for mu2 to MusicXML conversion (courtesy of Mogens Lundholm)

#### SymbTr-extras v0.3.2
 - Updated makammusicbrainz package version

#### SymbTr-extras v0.3.1
 - Updated required packages
 - Added rest row correction

#### SymbTr-extras v0.3.0
 - Improved code quality
 - Added unittests
 - Upgraded the required symbtrdataextractor to v2.0.0

#### SymbTr-extras v0.3.dev
 - Upgraded the required symbtrdataextractor to v2.0.0-alpha.5

#### SymbTr-extras v0.2.dev
 - Improved SymbTr MBID matching

#### SymbTr-extras v0.1
 - First public release


symbtrextras/unittests/test_extras.py
import os
from symbtrextras.scoreextras import ScoreExtras
from symbtrextras.txtextras import TxtExtras
import tempfile
import shutil
import filecmp


_curr_folder = os.path.dirname(os.path.abspath(__file__))


def test_wrong_score():
    run_all_extras('sedaraban--sazsemaisi--aksaksemai----tanburi_cemil_bey')


def run_all_extras(symbtr_name):
    # I/O, change the name to the desired SymbTr-score accordingly
    txt_file = os.path.join(_curr_folder, 'data', symbtr_name + '.txt')
    mu2_file = os.path.join(_curr_folder, 'data', symbtr_name + '.mu2')
    xml_file = os.path.join(_curr_folder, 'data', symbtr_name + '.xml')

    # create temporary files
    txt_temp = _create_temporary_copy(txt_file)
    mu2_temp = _create_temporary_copy(mu2_file)

    try:
        # Change encoding to UTF-8
        ScoreExtras.change_encoding_to_utf8(txt_temp)
        ScoreExtras.change_encoding_to_utf8(mu2_temp)

        # Change the line break from \r\n (Windows) to \n (Unix)
        ScoreExtras.change_to_unix_linebreak(txt_temp)
        ScoreExtras.change_to_unix_linebreak(mu2_temp)

        # check mu2
        mu2_corrected = os.path.join(_curr_folder, 'data', symbtr_name +
                                     '--corrected.mu2')
        assert filecmp.cmp(mu2_temp, mu2_corrected), \
            'Fault in mu2-extras: {0:s}'.format(symbtr_name)

        # Add usul row if it is not present in the first row
        symbtr_csv = TxtExtras.add_usul_to_first_row(txt_temp, mu2_temp)

        with open(txt_temp, 'w') as f:
            f.write(symbtr_csv)

        # Check usul rows and attempt to fill missing information
        symbtr_csv = TxtExtras.check_usul_row(txt_temp)

        with open(txt_temp, 'w') as text_file:
            text_file.write(symbtr_csv)

        # Force the duration of all gracenotes to 0 and recompute the offset
        # column according to the usul cycle(s)
        symbtr_csv = TxtExtras.correct_offset_gracenote(txt_temp, mu2_temp)
        with open(txt_temp, 'w') as text_file:
            text_file.write(symbtr_csv)

        # check txt
        txt_corrected = os.path.join(_curr_folder, 'data', symbtr_name +
                                     '--corrected.txt')
        assert filecmp.cmp(txt_temp, txt_corrected),\
            'Fault in txt-extras: {0:s}'.format(symbtr_name)

        # Convert the txt file to MusicXML using the additional metadata from
        # mu2 files (and the MusicBrainz mbid)
        symbtr_xml = TxtExtras.to_musicxml(symbtr_name, txt_temp, mu2_temp)

        # with open(xml_file, 'w') as xml_file:
        #     xml_file.write(symbtr_xml)

        with open(xml_file, 'r') as xml_file:
            saved_xml = xml_file.read()

        assert symbtr_xml == saved_xml, \
            'Fault in xml-conversion: {0:s}'.format(symbtr_name)
    finally:  # remove temporary files
        os.remove(txt_temp)
        os.remove(mu2_temp)


def _create_temporary_copy(path):
    temp_dir = tempfile.gettempdir()
    temp_path = os.path.join(temp_dir, os.path.basename(path))
    shutil.copy2(path, temp_path)
    return temp_path


symbtrextras/Extras.py
# -*- coding: utf-8 -*-
from symbtrdataextractor import symbtrreader, extractor
import pandas as pd 
import json

class Extras:
    def __init__(self):
        self.symbtr_cols = ['Sira', 'Kod', 'Nota53', 'NotaAE', 'Koma53', 'KomaAE', 
               'Pay', 'Payda', 'Ms', 'LNS', 'Bas', 'Soz1', 'Offset']
        self.usuldict = json.load(open('./symbtrextras/data/usul_extended.json','r'))

    def addUsul2FirstRow(self, txt_file, mu2_file):
        # extract symbtr data
        txtdata = extractor.extract(txt_file)[0]
        mu2header = symbtrreader.readMu2Header(mu2_file)[0]

        data = extractor.merge(txtdata, mu2header)  # merge

        # get usul variant
        variant = {}
        for vrt in self.usuldict[data['usul']['symbtr_slug']]['variants']:
            if vrt['mu2_name'] == data['usul']['mu2_name']:
                variant = vrt
                break

        # read the txt score
        df = pd.read_csv(txt_file, sep='\t')

        # create the usul row
        # 1    51            0    0    zaman    mertebe    0    usul_symbtr_internal_id    0    usul_mu2_name    0
        # 1    51            0    0    6    4    0    90    0    Yürüksemâî (6/4)    0
        usul_row = pd.DataFrame({'Sira':1, 'Kod':51, 'Nota53':'', 'NotaAE':'', 'Koma53':0, 'KomaAE':0,
            'Pay':int(variant['num_pulses']),'Payda':int(variant['mertebe']),'Ms':0,'Offset':0,
            'LNS':variant['symbtr_internal_id'],'Bas':0,'Soz1':variant['mu2_name']}, index=[0])

        if not df.iloc[0]['Kod'] == 51:            
            for index, row in df.iterrows():       
                # change null to empty string
                for key, val in row.iteritems():
                    if pd.isnull(val):
                        row[key] = ''

                # make sure that "Sira" column continues consecutively
                row['Sira'] = index + 2

                # reassign
                df.iloc[index] = row
                
            df_usul = pd.concat([usul_row, df], ignore_index=True)[self.symbtr_cols]
        else:
            if not df.iloc[0]["LNS"] == usul_row.iloc[0]["LNS"]:
                print data['symbtr'] + " starts with a different usul row. Correcting..."
                df_usul = pd.concat([usul_row, df.ix[1:]], ignore_index=True)[self.symbtr_cols]
            else:
                print data['symbtr'] + " starts with the usul row. Skipping..."
                df_usul = df

        return df_usul.to_csv(None, sep='\t', index=False, encoding='utf-8')

symbtrextras/TxtExtras.py
# -*- coding: utf-8 -*-
from __future__ import unicode_literals
from musicxmlconverter.symbtr2musicxml import SymbTrScore
from ScoreExtras import ScoreExtras
import pandas as pd
import os
import warnings
from future.utils import iteritems


class TxtExtras:
    symbtr_cols = ['Sira', 'Kod', 'Nota53', 'NotaAE', 'Koma53', 'KomaAE',
                   'Pay', 'Payda', 'Ms', 'LNS', 'Bas', 'Soz1', 'Offset']

    @classmethod
    def check_usul_row(cls, txt_file):
        mu2_usul_dict, inv_mu2_usul_dict = ScoreExtras.parse_usul_dict()

        df = pd.read_csv(txt_file, sep='\t', encoding='utf-8')

        symbtr_name = os.path.splitext(txt_file)[0]

        for index, row in df.iterrows():
            # change null to empty string
            row_changed = cls._change_null_element_to_empty_str(row)

            if row['Kod'] == 51:
                row_changed = cls._parse_usul_row(
                    row, index, mu2_usul_dict, inv_mu2_usul_dict, symbtr_name,
                    row_changed)

            if row_changed:
                df.iloc[index] = row

        return df.to_csv(None, sep=b'\t', index=False, encoding='utf-8')

    @classmethod
    def _parse_usul_row(cls, row, index, mu2_usul_dict, inv_mu2_usul_dict,
                        symbtr_name, row_changed):
        usul_id = row['LNS']
        usul_name = row['Soz1']
        if usul_name:  # name given
            cls._chk_usul_by_name(index, mu2_usul_dict, row, symbtr_name,
                                  usul_id, usul_name)
        elif usul_id:
            row_changed = cls._chk_usul_by_id(
                index, inv_mu2_usul_dict, row,
                symbtr_name, usul_id, usul_name, row_changed)
        else:
            raise RuntimeError("Unexpected operation")

        return row_changed

    @classmethod
    def _chk_usul_by_name(cls, index, mu2_usul_dict, row, symbtr_name, usul_id,
                          usul_name):
        # check if the usul pair matches with the mu2dict
        if usul_name in mu2_usul_dict and mu2_usul_dict[usul_name]['id'] == \
                usul_id:
            cls._chk_usul_attr(row, mu2_usul_dict[usul_name], 'zaman',
                               symbtr_name, index, usul_name)
            cls._chk_usul_attr(row, mu2_usul_dict[usul_name], 'mertebe',
                               symbtr_name, index, usul_name)
        else:
            warnstr = u'{0:s}, line {1:s}: {2:s} and {3:s} does not match.'.\
                format(symbtr_name, str(index), usul_name, str(usul_id))
            warnings.warn(warnstr.encode('utf-8'))

    @classmethod
    def _chk_usul_by_id(cls, index, inv_mu2_usul_dict, row, symbtr_name,
                        usul_id, usul_name, row_changed):
        if usul_id == -1:
            warnings.warn(u'{0:s}, line {1:s}: Missing usul info'.format(
                symbtr_name, str(index)).encode('utf-8'))
        else:
            warnstr = u'{0:s}, line {1:d}: Filling missing {2:s}'.format(
                symbtr_name, index, inv_mu2_usul_dict[usul_id]['mu2_name'])
            warnings.warn(warnstr.encode('utf-8'))
            row['Soz1'] = inv_mu2_usul_dict[usul_id]['mu2_name']

            cls._chk_usul_attr(row, inv_mu2_usul_dict[usul_id], 'zaman',
                               symbtr_name, index, usul_name)
            cls._chk_usul_attr(row, inv_mu2_usul_dict[usul_id], 'mertebe',
                               symbtr_name, index, usul_name)
            row_changed = True
        return row_changed

    @staticmethod
    def _change_null_element_to_empty_str(row):
        row_changed = False
        for key, val in row.iteritems():
            if pd.isnull(val):
                row[key] = ''
                row_changed = True
        return row_changed

    @staticmethod
    def _chk_usul_attr(row, usul, attr_str, symbtr_name, index, usul_name):
        if attr_str == 'mertebe':
            row_str = 'Payda'
        elif attr_str == 'zaman':
            row_str = 'Pay'
        else:
            raise ValueError(
                u'Unexpected usul attribute: {0:s}'.format(attr_str))
        if not usul[attr_str] == row[row_str]:
            warnstr = u'{0:s}, line {1:s}: {2:s} and {3:s} does not match.'\
                .format(symbtr_name, str(index), usul_name, attr_str)
            warnings.warn(warnstr.encode('utf-8'))

    @classmethod
    def add_usul_to_first_row(cls, txt_file, mu2_file):
        # extract symbtr data
        data = ScoreExtras.get_symbtr_data(txt_file, mu2_file)

        # get usul variant
        variant = cls._get_usul_variant(data)  # read the txt score
        df = pd.read_csv(txt_file, sep=b'\t')

        # create the usul row
        # 1    51            0    0    zaman    mertebe    0    usul_symbtr_internal_id    0    usul_mu2_name    0
        # 1    51            0    0    6    4    0    90    0    Yürüksemâî (6/4)    0
        usul_row = pd.DataFrame(
            {'Sira': 1, 'Kod': 51, 'Nota53': '', 'NotaAE': '', 'Koma53': 0,
             'KomaAE': 0, 'Pay': int(variant['num_pulses']),
             'Payda': int(variant['mertebe']), 'Ms': 0, 'Offset': 0,
             'LNS': variant['symbtr_internal_id'], 'Bas': 0,
             'Soz1': variant['mu2_name']}, index=[0])

        if not df.iloc[0]['Kod'] == 51:
            for index, row in df.iterrows():
                cls._change_null_to_empty_str(row)

                # make sure that "Sira" column continues consecutively
                row['Sira'] = index + 2  # 2 instead of 1, since we also add
                # the usul row to the start

                # reassign
                df.iloc[index] = row

            df_usul = pd.concat(
                [usul_row, df], ignore_index=True)[cls.symbtr_cols]
        else:
            if not df.iloc[0]["LNS"] == usul_row.iloc[0]["LNS"]:
                print(u"{0:s} starts with a different usul row. Correcting...".
                      format(data['symbtr']).encode('utf-8'))
                df_usul = pd.concat(
                    [usul_row, df.ix[1:]], ignore_index=True)[cls.symbtr_cols]
            else:
                print(u"{0:s} starts with the usul row. Skipping...".format(
                    data['symbtr']).encode('utf-8'))
                df_usul = df

        return df_usul.to_csv(None, sep=b'\t', index=False, encoding='utf-8')

    @classmethod
    def correct_offset_gracenote(cls, txt_file, mu2_file):
        # extract symbtr data
        data = ScoreExtras.get_symbtr_data(txt_file, mu2_file)

        # get zaman and mertebe from usul variant
        mertebe, zaman = cls._get_zaman_mertebe(data)

        # correct the offsets and the gracenote durations
        df = pd.read_csv(txt_file, sep=b'\t')
        row = dict()
        for index, row in df.iterrows():
            # recompute the erroneous gracenotes with non-zero duration
            if row['Kod'] == 8 and row['Ms'] > 0:
                row['Pay'] = 0
                row['Payda'] = 0
                row['Ms'] = 0

            # recompute zaman and mertebe, if we hit kod 51
            if row['Kod'] == 51:
                zaman = row['Pay']
                mertebe = row['Payda']
                offset_incr = 0
            else:
                # compute offset
                offset_incr = 0 if row['Payda'] == 0 else \
                    float(row['Pay']) / row['Payda'] * mertebe / zaman

            if index == 0:
                row['Offset'] = offset_incr
            else:
                prev_row = df.iloc[index - 1]
                row['Offset'] = offset_incr + prev_row['Offset']

            # change null to empty string
            cls._change_null_to_empty_str(row)

            # make sure that "Sira" column continues consecutively
            row['Sira'] = index + 1

            # reassign
            df.iloc[index] = row

        cls._check_premature_ending(row)

        return df.to_csv(None, sep=b'\t', index=False, encoding='utf-8')

    @staticmethod
    def _check_premature_ending(row):
        # warn if the last measure end prematurely, i.e. the last note does not
        # have an integer offset
        if not (round(row['Offset'] * 10000) * 0.0001).is_integer():
            warnings.warn("Ends prematurely!")

    @staticmethod
    def _get_usul_variant(data):
        usul_dict = ScoreExtras.get_usul_dict()
        vrts = usul_dict[data['usul']['symbtr_slug']]['variants']
        for v in vrts:
            if v['mu2_name'] == data['usul']['mu2_name']:
                return v

        assert False, u'The usul variant {0:s} is not found'.format(
            data['usul']['mu2_name'])

    @staticmethod
    def _get_zaman_mertebe(data):
        usul_dict = ScoreExtras.get_usul_dict()
        for usul in usul_dict.values():
            for uv in usul['variants']:
                if uv['mu2_name'] == data['usul']['mu2_name']:
                    return uv['mertebe'], uv['num_pulses']

        assert False, u'Zaman and mertebe for the usul variant {0:s} is not ' \
                      u'available'.format(data['usul']['mu2_name'])

    @staticmethod
    def to_musicxml(symbtr_name, txt_file, mu2_file):
        mbids = ScoreExtras.get_mbids(symbtr_name)

        # MusicXML conversion
        piece = SymbTrScore(txt_file, mu2_file, symbtrname=symbtr_name,
                            mbid_url=mbids)
        return piece.convertsymbtr2xml(verbose=False)

    @staticmethod
    def _change_null_to_empty_str(row):
        # change null to empty string
        for key, val in iteritems(row):
            if pd.isnull(val):
                row[key] = ''


symbtrextras/ScoreExtras.py
# -*- coding: utf-8 -*-
from __future__ import unicode_literals
import json
import urllib
import os
import subprocess
import warnings
from future.utils import iteritems
from symbtrdataextractor.SymbTrDataExtractor import SymbTrDataExtractor
from symbtrdataextractor.reader.Mu2Reader import Mu2Reader


class ScoreExtras:
    @staticmethod
    def _read_symbtr_mbid():
        try:
            url = "https://raw.githubusercontent.com/MTG/SymbTr/master/" \
                  "symbTr_mbid.json"
            response = urllib.urlopen(url)
            return json.loads(response.read())
        except IOError:  # if it is called from the submodule location or it
            # was installed with "pip install -e ." we can refer to the json
            # in the local repo
            try:
                warnings.warn("Cannot access the symbtr_mbid.json in the "
                              "github SymbTr repository. Falling back to the "
                              "json file in the local SymbTr repository.")
                return json.load(open(  # load symbTr mbids
                    os.path.join(os.path.dirname(os.path.abspath(__file__)),
                                 '..', '..', 'symbTr_mbid.json'), 'r'))
            except IOError:
                warnings.warn("symbtr_mbid.json is not found in the local "
                              "search path. Using the back-up "
                              "symbtr_mbid.json included in this repository.")
                return ScoreExtras.load_local_json('symbTr_mbid.json')

    @staticmethod
    def get_usul_dict():
        return ScoreExtras.load_local_json('usul_extended.json')

    @staticmethod
    def load_local_json(json_name):
        return json.load(open(  # load symbTr mbids
            os.path.join(os.path.dirname(os.path.abspath(__file__)),
                         'data', json_name), 'r'))

    _iconv_map = {'utf-16le': 'UTF-16',
                  'Little-endian UTF-16 Unicode': 'UTF-16',
                  'iso-8859-1': 'ISO_8859-9', 'ISO-8859': 'ISO_8859-9'}

    @staticmethod
    def get_symbtr_data(txt_file, mu2_file):
        extractor = SymbTrDataExtractor()
        txt_data = extractor.extract(txt_file)[0]

        mu2_header = Mu2Reader.read_header(mu2_file)[0]

        return extractor.merge(txt_data, mu2_header, verbose=False)

    @classmethod
    def get_mbids(cls, symbtr_name):
        mbids = []  # extremely rare but there can be more than one mbid
        for e in cls._read_symbtr_mbid():
            if e['name'] == symbtr_name:
                mbids.append(e['uuid'])
        return mbids

    @classmethod
    def parse_usul_dict(cls):
        mu2_usul_dict = {}
        inv_mu2_usul_dict = {}
        usul_dict = cls.get_usul_dict()
        for key, val in iteritems(usul_dict):
            for vrt in val['variants']:
                if vrt['mu2_name']:  # if it doesn't have a mu2 name, the usul
                    # is not in symbtr collection
                    zaman = int(vrt['num_pulses']) if vrt['num_pulses'] else []
                    mertebe = int(vrt['mertebe']) if vrt['mertebe'] else []
                    if vrt['mu2_name'] in ['(Serbest)', '[Serbest]',
                                           'Serbest']:
                        zaman = 0
                        mertebe = 0
                    mu2_usul_dict[vrt['mu2_name']] = {
                        'id': int(vrt['symbtr_internal_id']), 'zaman': zaman,
                        'mertebe': mertebe}

                    inv_mu2_usul_dict[int(vrt['symbtr_internal_id'])] = {
                        'mu2_name': vrt['mu2_name'], 'zaman': zaman,
                        'mertebe': mertebe}
        return mu2_usul_dict, inv_mu2_usul_dict

    @staticmethod
    def change_to_unix_linebreak(score_file):
        # change the line break from \r\n to \n
        try:  # linux
            subprocess.check_call("fromdos " + score_file, shell=True)
        except subprocess.CalledProcessError:  # mac
            subprocess.check_call("sed -e 's/\r$//' " + score_file +
                                  " > tmp.txt " + "&& mv -f tmp.txt " +
                                  score_file, shell=True)

    @classmethod
    def change_encoding_to_utf8(cls, score_file):
        try:  # unix
            out = subprocess.check_output("file -i " + score_file, shell=True)
            curr_charset = out.split('charset=')[1]

            if curr_charset.endswith('\n'):
                curr_charset = curr_charset[:-1]

            if not any(curr_charset in charset for charset in ['utf-8',
                                                               'us-ascii']):
                print(curr_charset + '\t' + score_file)
                commandstr = ("iconv -f " + cls._iconv_map[curr_charset] +
                              " -t UTF-8 " + score_file + " > tmp.txt "
                              "&& mv -f tmp.txt " + score_file)
                subprocess.check_output(commandstr, shell=True)
        except IndexError:  # mac
            raise OSError('Call this method in Linux for reliable results.')


symbtrextras/txtextras.py
# -*- coding: utf-8 -*-
from __future__ import unicode_literals
from musicxmlconverter.symbtr2musicxml import SymbTrScore
from scoreextras import ScoreExtras
import pandas as pd
import os
import warnings
from future.utils import iteritems


class TxtExtras:
    symbtr_cols = ['Sira', 'Kod', 'Nota53', 'NotaAE', 'Koma53', 'KomaAE',
                   'Pay', 'Payda', 'Ms', 'LNS', 'Bas', 'Soz1', 'Offset']

    @classmethod
    def check_usul_row(cls, txt_file):
        mu2_usul_dict, inv_mu2_usul_dict = ScoreExtras.parse_usul_dict()

        df = pd.read_csv(txt_file, sep='\t', encoding='utf-8')

        symbtr_name = os.path.splitext(txt_file)[0]

        for index, row in df.iterrows():
            # change null to empty string
            row_changed = cls._change_null_element_to_empty_str(row)

            if row['Kod'] == 51:
                row_changed = cls._parse_usul_row(
                    row, index, mu2_usul_dict, inv_mu2_usul_dict, symbtr_name,
                    row_changed)

            if row_changed:
                df.iloc[index] = row

        return df.to_csv(None, sep=b'\t', index=False, encoding='utf-8')

    @classmethod
    def _parse_usul_row(cls, row, index, mu2_usul_dict, inv_mu2_usul_dict,
                        symbtr_name, row_changed):
        usul_id = row['LNS']
        usul_name = row['Soz1']
        if usul_name:  # name given
            cls._chk_usul_by_name(index, mu2_usul_dict, row, symbtr_name,
                                  usul_id, usul_name)
        elif usul_id:
            row_changed = cls._chk_usul_by_id(
                index, inv_mu2_usul_dict, row,
                symbtr_name, usul_id, usul_name, row_changed)
        else:
            raise RuntimeError("Unexpected operation")

        return row_changed

    @classmethod
    def _chk_usul_by_name(cls, index, mu2_usul_dict, row, symbtr_name, usul_id,
                          usul_name):
        # check if the usul pair matches with the mu2dict
        if usul_name in mu2_usul_dict and mu2_usul_dict[usul_name]['id'] == \
                usul_id:
            cls._chk_usul_attr(row, mu2_usul_dict[usul_name], 'zaman',
                               symbtr_name, index, usul_name)
            cls._chk_usul_attr(row, mu2_usul_dict[usul_name], 'mertebe',
                               symbtr_name, index, usul_name)
        else:
            warnstr = u'{0:s}, line {1:s}: {2:s} and {3:s} does not match.'.\
                format(symbtr_name, str(index), usul_name, str(usul_id))
            warnings.warn(warnstr.encode('utf-8'))

    @classmethod
    def _chk_usul_by_id(cls, index, inv_mu2_usul_dict, row, symbtr_name,
                        usul_id, usul_name, row_changed):
        if usul_id == -1:
            warnings.warn(u'{0:s}, line {1:s}: Missing usul info'.format(
                symbtr_name, str(index)).encode('utf-8'))
        else:
            warnstr = u'{0:s}, line {1:d}: Filling missing {2:s}'.format(
                symbtr_name, index, inv_mu2_usul_dict[usul_id]['mu2_name'])
            warnings.warn(warnstr.encode('utf-8'))
            row['Soz1'] = inv_mu2_usul_dict[usul_id]['mu2_name']

            cls._chk_usul_attr(row, inv_mu2_usul_dict[usul_id], 'zaman',
                               symbtr_name, index, usul_name)
            cls._chk_usul_attr(row, inv_mu2_usul_dict[usul_id], 'mertebe',
                               symbtr_name, index, usul_name)
            row_changed = True
        return row_changed

    @staticmethod
    def _change_null_element_to_empty_str(row):
        row_changed = False
        for key, val in row.iteritems():
            if pd.isnull(val):
                row[key] = ''
                row_changed = True
        return row_changed

    @staticmethod
    def _chk_usul_attr(row, usul, attr_str, symbtr_name, index, usul_name):
        if attr_str == 'mertebe':
            row_str = 'Payda'
        elif attr_str == 'zaman':
            row_str = 'Pay'
        else:
            raise ValueError(
                u'Unexpected usul attribute: {0:s}'.format(attr_str))
        if not usul[attr_str] == row[row_str]:
            warnstr = u'{0:s}, line {1:s}: {2:s} and {3:s} does not match.'\
                .format(symbtr_name, str(index), usul_name, attr_str)
            warnings.warn(warnstr.encode('utf-8'))

    @classmethod
    def add_usul_to_first_row(cls, txt_file, mu2_file):
        # extract symbtr data
        data = ScoreExtras.get_symbtr_data(txt_file, mu2_file)

        # get usul variant
        variant = cls._get_usul_variant(data)  # read the txt score
        df = pd.read_csv(txt_file, sep=b'\t')

        # create the usul row
        # 1    51            0    0    zaman    mertebe    0    usul_symbtr_internal_id    0    usul_mu2_name    0
        # 1    51            0    0    6    4    0    90    0    Yürüksemâî (6/4)    0
        usul_row = pd.DataFrame(
            {'Sira': 1, 'Kod': 51, 'Nota53': '', 'NotaAE': '', 'Koma53': 0,
             'KomaAE': 0, 'Pay': int(variant['num_pulses']),
             'Payda': int(variant['mertebe']), 'Ms': 0, 'Offset': 0,
             'LNS': variant['symbtr_internal_id'], 'Bas': 0,
             'Soz1': variant['mu2_name']}, index=[0])

        if not df.iloc[0]['Kod'] == 51:
            for index, row in df.iterrows():
                cls._change_null_to_empty_str(row)

                # make sure that "Sira" column continues consecutively
                row['Sira'] = index + 2  # 2 instead of 1, since we also add
                # the usul row to the start

                # reassign
                df.iloc[index] = row

            df_usul = pd.concat(
                [usul_row, df], ignore_index=True)[cls.symbtr_cols]
        else:
            if not df.iloc[0]["LNS"] == usul_row.iloc[0]["LNS"]:
                print(u"{0:s} starts with a different usul row. Correcting...".
                      format(data['symbtr']).encode('utf-8'))
                df_usul = pd.concat(
                    [usul_row, df.ix[1:]], ignore_index=True)[cls.symbtr_cols]
            else:
                print(u"{0:s} starts with the usul row. Skipping...".format(
                    data['symbtr']).encode('utf-8'))
                df_usul = df

        return df_usul.to_csv(None, sep=b'\t', index=False, encoding='utf-8')

    @classmethod
    def correct_offset_gracenote(cls, txt_file, mu2_file):
        # extract symbtr data
        data = ScoreExtras.get_symbtr_data(txt_file, mu2_file)

        # get zaman and mertebe from usul variant
        mertebe, zaman = cls._get_zaman_mertebe(data)

        # correct the offsets and the gracenote durations
        df = pd.read_csv(txt_file, sep=b'\t')
        for index, row in df.iterrows():
            # recompute the erroneous gracenotes with non-zero duration
            if row['Kod'] == 8 and row['Ms'] > 0:
                row['Pay'] = 0
                row['Payda'] = 0
                row['Ms'] = 0

            # recompute zaman and mertebe, if we hit kod 51
            if row['Kod'] == 51:
                zaman = row['Pay']
                mertebe = row['Payda']
                offset_incr = 0
            else:
                # compute offset
                offset_incr = 0 if row['Payda'] == 0 else \
                    float(row['Pay']) / row['Payda'] * mertebe / zaman

            if index == 0:
                row['Offset'] = offset_incr
            else:
                prev_row = df.iloc[index - 1]
                row['Offset'] = offset_incr + prev_row['Offset']

            # change null to empty string
            cls._change_null_to_empty_str(row)

            # make sure that "Sira" column continues consecutively
            row['Sira'] = index + 1

            # reassign
            df.iloc[index] = row

        cls._check_premature_ending(row)

        return df.to_csv(None, sep=b'\t', index=False, encoding='utf-8')

    @staticmethod
    def correct_rests(txt_file):
        # correct the offsets and the gracenote durations
        df = pd.read_csv(txt_file, sep=b'\t')
        rest_list = [9, -1, -1, 'Es', 'Es']
        for index, row in df.iterrows():
            val_list = [row['Kod'], row['Koma53'], row['KomaAE'],
                        row['Nota53'], row['NotaAE']]

            # check if rest. If yes, check if valid
            is_rest = any(v1 == v2 for v1, v2 in
                          zip(val_list[1:], rest_list[1:]))
            is_invalid = any(v1 != v2 for v1, v2 in zip(val_list, rest_list))
            if is_rest and is_invalid:
                (row['Kod'], row['Koma53'], row['KomaAE'], row['Nota53'],
                 row['NotaAE']) = rest_list
                df.iloc[index] = row
        return df.to_csv(None, sep=b'\t', index=False, encoding='utf-8')

    @staticmethod
    def _check_premature_ending(row):
        # warn if the last measure end prematurely, i.e. the last note does not
        # have an integer offset
        if not (round(row['Offset'] * 10000) * 0.0001).is_integer():
            warnings.warn("Ends prematurely!")

    @staticmethod
    def _get_usul_variant(data):
        usul_dict = ScoreExtras.get_usul_dict()
        vrts = usul_dict[data['usul']['symbtr_slug']]['variants']
        for v in vrts:
            if v['mu2_name'] == data['usul']['mu2_name']:
                return v

        assert False, u'The usul variant {0:s} is not found'.format(
            data['usul']['mu2_name'])

    @staticmethod
    def _get_zaman_mertebe(data):
        usul_dict = ScoreExtras.get_usul_dict()
        for usul in usul_dict.values():
            for uv in usul['variants']:
                if uv['mu2_name'] == data['usul']['mu2_name']:
                    return uv['mertebe'], uv['num_pulses']

        assert False, u'Zaman and mertebe for the usul variant {0:s} is not ' \
                      u'available'.format(data['usul']['mu2_name'])

    @staticmethod
    def to_musicxml(symbtr_name, txt_file, mu2_file):
        mbids = ScoreExtras.get_mbids(symbtr_name)

        # MusicXML conversion
        piece = SymbTrScore(txt_file, mu2_file, symbtrname=symbtr_name,
                            mbid_url=mbids, verbose=True)

        return piece.convertsymbtr2xml(verbose=False)

    @staticmethod
    def _change_null_to_empty_str(row):
        # change null to empty string
        for key, val in iteritems(row):
            if pd.isnull(val):
                row[key] = ''
